### PR TITLE
admin: add KMS status API

### PR DESCRIPTION
**Pull request by aead**
## Description
This commit adds an admin API for fetching
the KMS status information (default key ID, endpoints, ...).

With this commit the server exposes REST endpoint:
```
GET <admin-api>/kms/status
```

## Motivation and Context
KMS, Currently WIP - depends on minio/madmin-go#10

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#3251e8a5-4bc9-433c-b31f-c38ecb7c653f) |
<a name="3251e8a5-4bc9-433c-b31f-c38ecb7c653f"></a>
### 12429-8071423/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:31723
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 3a261f2588b7:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... done in 2 minutes and 15 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 1 minutes and 18 seconds
(8/15) Running minio-dotnet tests ... done in 43 seconds
(9/15) Running minio-go tests ... done in 2 minutes and 11 seconds
(10/15) Running minio-java tests ... FAILED in 1 minutes and 24 seconds
{
  "name": "minio-java",
  "function": "composeObject()",
  "args": "[single source with offset]",
  "duration": 29,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = InvalidArgument, message = Range specified is not valid for source object, bucketName = minio-java-test-234eqf5, objectName = minio-java-test-1qladb4, resource = /minio-java-test-234eqf5/minio-java-test-1qladb4, requestId = 16865984A8BB61FA, hostId = a5584bbc-6a18-47c3-9c37-b6ff2c19b682)\nrequest={method=PUT, url=http://minio-c3.minio.io:31723/minio-java-test-234eqf5/minio-java-test-1qladb4?uploadId=596c494e-f84c-438a-8a41-eebd1c2870cb&partNumber=1, headers=x-amz-copy-source: /minio-java-test-234eqf5/minio-java-test-k7lmhr\nx-amz-copy-source-range: bytes=2048-1048575\nx-amz-copy-source-if-match: cb92d17a904ccec2e6e23b8bb66245fb\nHost: minio-c3.minio.io:31723\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210607T162542Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210607/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-copy-source-range;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=400, headers=Accept-Ranges: bytes\nContent-Length: 390\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 16865984A8BB61FA\nX-Xss-Protection: 1; mode=block\nDate: Mon, 07 Jun 2021 16:25:42 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.uploadPartCopy(MinioClient.java:4804), io.minio.MinioClient.composeObject(MinioClient.java:1431), FunctionalTest.testComposeObject(FunctionalTest.java:2120), FunctionalTest.composeObjectTests(FunctionalTest.java:2145), FunctionalTest.composeObject(FunctionalTest.java:2300), FunctionalTest.runObjectTests(FunctionalTest.java:3758), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(10/15) Running minio-js tests ... done in 48 seconds
(11/15) Running minio-py tests ... done in 2 minutes and 52 seconds
(12/15) Running s3cmd tests ... done in 19 seconds
(13/15) Running s3select tests ... done in 7 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#df777bb5-6ca6-42b8-90b8-e46c80d20c17) |
<a name="df777bb5-6ca6-42b8-90b8-e46c80d20c17"></a>
### 12429-7763cc7/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:32673
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 30ad798ac919:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... done in 2 minutes and 12 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 1 minutes and 8 seconds
(8/15) Running minio-dotnet tests ... done in 46 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 52 seconds
(10/15) Running minio-java tests ... FAILED in 1 minutes and 27 seconds
{
  "name": "minio-java",
  "function": "composeObject()",
  "args": "[single source with offset]",
  "duration": 28,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = InvalidArgument, message = Range specified is not valid for source object, bucketName = minio-java-test-vp51ro, objectName = minio-java-test-2sj9a2n, resource = /minio-java-test-vp51ro/minio-java-test-2sj9a2n, requestId = 16865A8BE57808EE, hostId = 158e763b-b749-483f-84bb-10aefbc25f51)\nrequest={method=PUT, url=http://minio-c2.minio.io:32673/minio-java-test-vp51ro/minio-java-test-2sj9a2n?uploadId=57ce23b7-2fa2-4b10-93be-2d60eb591ee4&partNumber=1, headers=x-amz-copy-source: /minio-java-test-vp51ro/minio-java-test-oihql0\nx-amz-copy-source-range: bytes=2048-1048575\nx-amz-copy-source-if-match: cb92d17a904ccec2e6e23b8bb66245fb\nHost: minio-c2.minio.io:32673\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210607T164432Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210607/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-copy-source-range;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=400, headers=Accept-Ranges: bytes\nContent-Length: 388\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 16865A8BE57808EE\nX-Xss-Protection: 1; mode=block\nDate: Mon, 07 Jun 2021 16:44:32 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.uploadPartCopy(MinioClient.java:4804), io.minio.MinioClient.composeObject(MinioClient.java:1431), FunctionalTest.testComposeObject(FunctionalTest.java:2120), FunctionalTest.composeObjectTests(FunctionalTest.java:2145), FunctionalTest.composeObject(FunctionalTest.java:2300), FunctionalTest.runObjectTests(FunctionalTest.java:3758), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(10/15) Running minio-js tests ... done in 46 seconds
(11/15) Running minio-py tests ... done in 2 minutes and 59 seconds
(12/15) Running s3cmd tests ... done in 19 seconds
(13/15) Running s3select tests ... done in 6 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/admin-router.go [hunk 1]
```diff
@@ -201,6 +201,7 @@ func registerAdminRouter(router *mux.Router, enableConfigOps, enableIAMOps bool)
 
 		// -- KMS APIs --
 		//
+		adminRouter.Methods(http.MethodPost).Path(adminVersion + "/kms/status").HandlerFunc(httpTraceAll(adminAPI.KMSStatusHandler))
```
Comment (harshavardhana) [inline cmd/admin-router.go:204]: What is the difference between kms/key/status and kms/status? @aead

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/admin-router.go (hunk 1 above)
Comment (aead) [reply cmd/admin-router.go:204]: The `kms/status` API returns general information about the KMS - its endpoint(s), the default KMS key used for SSE-S3 and SSE-KMS when no explicit KMS key id is specified by a user / no SSE-KMS bucket configuration is set, etc.

The `kms/key/status` returns information about a particular key - like whether you can encrypt and decrypt.

Comment (aead) [reply cmd/admin-router.go:204]: From the `mc` side think of it as:
```
$ mc kms status <alias>
{
    "DefaultKeyID": "my-minio-key",
    "Endpoints:": {
          "https://kes1.local": "online",
     },
     ...
}

$ mc kms key create <alias> my-key

$ mc kms key status <alias> my-key
my-key:
   - Encrypt: ✓
   - Decrypt: ✓
```

Comment (harshavardhana) [reply cmd/admin-router.go:204]: Understood @aead is this similar to `mc admin info` is doing `fetchKMSStatus()` ?

Comment (aead) [reply cmd/admin-router.go:204]: Yeah - basically KMS-specific and with more user-friendly output.

Comment (harshavardhana) [reply cmd/admin-router.go:204]: Can your function be re-purposed for `mc admin info` - fetchKMSStatus() ?

Comment (aead) [reply cmd/admin-router.go:204]: Sure, that can be done 👍

**Review by harshavardhana — APPROVED** [review]

**Review by Alevsk — APPROVED** [review]

**Pull request merged**